src/MEVCapturingHook.sol

```diff
@@ -3,36 +3,102 @@ pragma solidity ^0.8.26;
 
 import {BaseHook} from "v4-periphery/src/utils/BaseHook.sol";
 
-// import {CurrencyLibrary, Currency} from "v4-core/types/Currency.sol";
-// import {PoolKey} from "v4-core/types/PoolKey.sol";
-// import {BalanceDeltaLibrary, BalanceDelta} from "v4-core/types/BalanceDelta.sol";
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+
+import {CurrencyLibrary, Currency} from "v4-core/types/Currency.sol";
+import {PoolKey} from "v4-core/types/PoolKey.sol";
+import {PoolId, PoolIdLibrary} from "v4-core/types/PoolId.sol";
+import {BalanceDeltaLibrary, BalanceDelta} from "v4-core/types/BalanceDelta.sol";
+import {toBeforeSwapDelta, BeforeSwapDelta, BeforeSwapDeltaLibrary} from "v4-core/types/BeforeSwapDelta.sol";
+import {LPFeeLibrary} from "v4-core/libraries/LPFeeLibrary.sol";
+import {Hooks} from "v4-core/libraries/Hooks.sol";
 
 import {IPoolManager} from "v4-core/interfaces/IPoolManager.sol";
 
-import {Hooks} from "v4-core/libraries/Hooks.sol";
+struct PoolConfig {
+    uint256 feeUnit;
+    uint256 priorityThreshold;
+}
+
+contract MEVCapturingHook is BaseHook, Ownable {
+    using LPFeeLibrary for uint24;
+    using PoolIdLibrary for PoolKey;
+
+    error MustUseDynamicFee();
+
+    uint256 public constant DEFAULT_FEE_UNIT = 1 wei; // ?? this is too low
+    uint256 public constant DEFAULT_PRIORITY_THRESHOLD = 10 wei; // ?? this is too low
+
+    mapping(PoolId => PoolConfig) poolConfig;
+    mapping(PoolId => uint256) lastTradedBlock;
 
-contract MEVCapturingHook is BaseHook {
     // Initialize BaseHook and ERC20
-    constructor(IPoolManager _manager) BaseHook(_manager) {}
+    constructor(IPoolManager _manager, address _initialOwner) BaseHook(_manager) Ownable(_initialOwner) {}
 
-    // Set up hook permissions to return `true`
-    // for the two hook functions we are using
     function getHookPermissions() public pure override returns (Hooks.Permissions memory) {
         return Hooks.Permissions({
-            beforeInitialize: false,
+            beforeInitialize: true,
             afterInitialize: false,
             beforeAddLiquidity: false,
             beforeRemoveLiquidity: false,
             afterAddLiquidity: false,
             afterRemoveLiquidity: false,
-            beforeSwap: false,
+            beforeSwap: true,
             afterSwap: false,
             beforeDonate: false,
             afterDonate: false,
-            beforeSwapReturnDelta: false,
+            beforeSwapReturnDelta: true,
             afterSwapReturnDelta: false,
             afterAddLiquidityReturnDelta: false,
             afterRemoveLiquidityReturnDelta: false
         });
     }
+
+    function _beforeInitialize(address, PoolKey calldata key, uint160) internal override returns (bytes4) {
+        if (!key.fee.isDynamicFee()) revert MustUseDynamicFee();
+
+        poolConfig[key.toId()] = PoolConfig(DEFAULT_FEE_UNIT, DEFAULT_PRIORITY_THRESHOLD);
```

Comment:
Should we use different default values depending on the decimal places for the token?

Reply:
yeah i think that probably makes sense
actually do we need 2 values? if token0 has 18 decimals and token1 has 6 we want different fee units depending on the direction of the swap right?

Reply:
yeah, makes sense...

Or alternatively we could also read before each swap value the decimal places of the input token and then scale DEFAULT_FEE_UNIT with it, right? Maybe this is the easiest...
We should also check that the currency wrapper is not doing that already...

Reply:
wdyt about the alternative, ie setting dynamic swap fee in percent via pool manager?
i think then we never run into the decimals problem right?